File: problems/binary_trees/compare_leaf_traversal/js/compare_leaf_traversal.js
export const compareLeafTraversal = (() => {
  const collectLeafNodes = (root, storedLeaf = []) => {
    if (!root) return storedLeaf;

    if (!root.left && !root.right) {
      storedLeaf.push(root.value);
    }
    collectLeafNodes(root.left, storedLeaf);
    collectLeafNodes(root.right, storedLeaf);
    return storedLeaf;
  };

  return (tree1, tree2) => {
    const leafTree1 = collectLeafNodes(tree1);
    const leafTree2 = collectLeafNodes(tree2);

    if (leafTree1.length !== leafTree2.length) return false;

    for (let i = 0; i < leafTree1.length; i++) {
      if (leafTree1[i] !== leafTree2[i]) return false;
    }
    return true;
  };
})();

File: problems/binary_trees/compare_leaf_traversal/python/compare_leaf_traversal.py
from problems.binary_trees.assets.BinaryTree import BinaryTree


def compareLeafTraversal(tree1: BinaryTree, tree2: BinaryTree) -> bool:

    def collect_leaf_nodes(root: BinaryTree, stored_leaf: list = None) -> list:
        if stored_leaf is None:
            stored_leaf = []

        if root is None:
            return stored_leaf

        if root.left is None and root.right is None:
            stored_leaf.append(root.value)

        collect_leaf_nodes(root.left, stored_leaf)
        collect_leaf_nodes(root.right, stored_leaf)

        return stored_leaf

    leaf_tree1 = collect_leaf_nodes(tree1)
    leaf_tree2 = collect_leaf_nodes(tree2)

    return leaf_tree1 == leaf_tree2
